lahja/base.py

```diff
@@ -211,6 +202,72 @@ def broadcast_nowait(
         """
         ...
 
+    #
+    # Subscription API
+    #
+    @abstractmethod
+    def get_connected_endpoints_and_subscriptions(
+        self
+    ) -> Tuple[Tuple[Optional[str], Set[Type[BaseEvent]]], ...]:
+        """
+        Return all connected endpoints and their event type subscriptions to this endpoint.
+        """
+        ...
+
+    @abstractmethod
+    def is_remote_subscribed_to(
+        self, remote_endpoint: str, event_type: Type[BaseEvent]
+    ) -> bool:
+        """
+        Return ``True`` if the specified remote endpoint is subscribed to the specified event type
+        from this endpoint. Otherwise return ``False``.
+        """
+        ...
+
+    @abstractmethod
+    def is_any_remote_subscribed_to(self, event_type: Type[BaseEvent]) -> bool:
+        """
+        Return ``True`` if at least one of the connected remote endpoints is subscribed to the
+        specified event type from this endpoint. Otherwise return ``False``.
+        """
+        ...
+
+    @abstractmethod
+    def are_all_remotes_subscribed_to(self, event_type: Type[BaseEvent]) -> bool:
+        """
+        Return ``True`` if every connected remote endpoint is subscribed to the specified event
+        type from this endpoint. Otherwise return ``False``.
+        """
+        ...
+
+    @abstractmethod
+    async def wait_until_remote_subscribed_to(
+        self, remote_endpoint: str, event: Type[BaseEvent]
+    ) -> None:
+        """
+        Block until the specified remote endpoint has subscribed to the specified event type
+        from this endpoint.
+        """
+        ...
+
+    @abstractmethod
+    async def wait_until_any_remote_subscribed_to(self, event: Type[BaseEvent]) -> None:
+        """
+        Block until any other remote endpoint has subscribed to the specified event type
+        from this endpoint.
+        """
+        ...
+
+    @abstractmethod
+    async def wait_until_all_remotes_subscribed_to(
+        self, event: Type[BaseEvent]
+    ) -> None:
```

Comment:
nitpick: Not part of this PR but these methods would be slightly easier to use if they included timeouts, you probably never want to call this method without wrapping it in something like [`wait_for`](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for).

Reply:
@lithp I've been coming around to the `trio` way of doing things which is that `timeout` parameters should not be part of the API but a responsibility of the caller.  

4th paragraph here: https://trio.readthedocs.io/en/latest/reference-core.html#handling-cancellation

I know that this concept doesn't map perfectly onto `asyncio` since asyncio doesn't have as nice of timeout support... but I still think there might be a nugget of wisdom in there.

Reply:
Nice link, this makes me even more excited to try working with trio. I'm definitely buying the argument that timeouts are more like exceptions; the code in the best position to determine the right timeout or exception handling strategy is usually relatively far away from the code which might generate those timeouts or exceptions.